sql/settings.txt
INSERT INTO `arma_myref_config` VALUES (1, 'version', '', '1', '', 'hidden', 'hidden', 0);
INSERT INTO `arma_myref_config` VALUES (2, 'lastraz', '', '32', '', 'hidden', 'hidden', 0);
INSERT INTO `arma_myref_config` VALUES (3, 'count_admin', '_MYREFERER_COUNT_ADMIN', '0', '_MYREFERER_COUNT_ADMIN_DSC', 'yesno', 'int', 0);
INSERT INTO `arma_myref_config` VALUES (4, 'module_blacklist', '_MYREFERER_MODULE_BLACKLIST', 'xoopsorgnews|xoops_redirect|user|search|notifications|viewpmsg|readpmsg|myreferer|newbb|system', '_MYREFERER_MODULE_BLACKLIST_DSC', 'textarea', 'text', 2);
INSERT INTO `arma_myref_config` VALUES (5, '', '_MYREFERER_KEYWORD', '', '', 'insertBreak', 'hidden', 100);
INSERT INTO `arma_myref_config` VALUES (6, 'keyword_min', '_MYREFERER_KEYWORD_MIN', '3', '_MYREFERER_KEYWORD_MIN_DSC', 'text', 'int', 101);
INSERT INTO `arma_myref_config` VALUES (7, 'keyword_max', '_MYREFERER_KEYWORD_MAX', '24', '_MYREFERER_KEYWORD_MAX_DSC', 'text', 'int', 102);
INSERT INTO `arma_myref_config` VALUES (8, 'punctation', '_MYREFERER_PUNCTATION', '1', '_MYREFERER_PUNCTATION_DSC', 'yesno', 'int', 103);
INSERT INTO `arma_myref_config` VALUES (9, 'numbers', '_MYREFERER_NUMBERS', '0', '_MYREFERER_NUMBERS_DSC', 'yesno', 'int', 104);
INSERT INTO `arma_myref_config` VALUES (10, 'smallcaps', '_MYREFERER_SMALLCAPS', '1', '_MYREFERER_SMALLCAPS_DSC', 'yesno', 'int', 105);
INSERT INTO `arma_myref_config` VALUES (11, 'keyword_blacklist', '_MYREFERER_KEYWORD_BLACKLIST', 'pour|avec|dans|site', '_MYREFERER_KEYWORD_BLACKLIST_DSC', 'textarea', 'text', 106);
INSERT INTO `arma_myref_config` VALUES (12, 'search_blacklist', '_MYREFERER_SEARCH_BLACKLIST', '', '_MYREFERER_SEARCH_BLACKLIST_DSC', 'textarea', 'text', 107);
INSERT INTO `arma_myref_config` VALUES (13, '', '_MYREFERER_REFERER', '', '', 'insertBreak', 'hidden', 200);
INSERT INTO `arma_myref_config` VALUES (14, 'referer_blacklist', '_MYREFERER_REFERER_BLACKLIST', '127.0.0.1|mail|xoopsorgnews', '_MYREFERER_REFERER_BLACKLIST_DSC', 'textarea', 'text', 201);
INSERT INTO `arma_myref_config` VALUES (15, '', '_MYREFERER_ROBOTS', '', '', 'insertBreak', 'hidden', 300);
INSERT INTO `arma_myref_config` VALUES (16, 'new_bot_smail', '_MYREFERER_NEW_BOT_SMAIL', '1', '_MYREFERER_NEW_BOT_SMAIL_DSC', 'yesno', 'int', 301);
INSERT INTO `arma_myref_config` VALUES (17, 'new_bot_mail', '_MYREFERER_NEW_BOT_MAIL', '', '_MYREFERER_NEW_BOT_MAIL_DSC', 'text', 'text', 302);
INSERT INTO `arma_myref_config` VALUES (18, 'robots_blacklist', '_MYREFERER_ROBOTS_BLACKLIST', 'W3C_Validator|Lynx|libwww-perl', '_MYREFERER_ROBOTS_BLACKLIST_DSC', 'textarea', 'text', 303);
INSERT INTO `arma_myref_config` VALUES (19, 'page_prohibit', '_MYREFERER_PAGE_PROHIBIT', 'modules/myreferer/norobot.html', '_MYREFERER_PAGE_PROHIBIT_DSC', 'text', 'text', 304);
INSERT INTO `arma_myref_config` VALUES (20, 'robots_prohibit', '_MYREFERER_ROBOTS_PROHIBIT', '', '_MYREFERER_ROBOTS_PROHIBIT_DSC', 'textarea', 'text', 305);
INSERT INTO `arma_myref_config` VALUES (21, '', '_MYREFERER_USERVISIT', '', '', 'insertBreak', 'hidden', 100);
INSERT INTO `arma_myref_config` VALUES (22, 'user_visit', '_MYREFERER_USER_VISIT', '1', '_MYREFERER_USER_VISIT_DSC', 'yesno', 'int', 401);
INSERT INTO `arma_myref_config` VALUES (23, 'save_group', '_MYREFERER_SAVE_GROUP', 'a:2:{i:0;s:1:"1";i:1;s:1:"2";}', '_MYREFERER_SAVE_GROUP_DSC', 'select', 'array', 402);
INSERT INTO `arma_myref_config` VALUES (24, '', '_MYREFERER_STATS', '', '', 'insertBreak', 'hidden', 400);
INSERT INTO `arma_myref_config` VALUES (25, 'myref_pages_stats', '_MYREFERER_STATS_PAGES', '100', '_MYREFERER_STATS_PAGES_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (26, 'myref_query_stats', '_MYREFERER_STATS_QUERY', '100', '_MYREFERER_STATS_QUERY_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (27, 'myref_query_pages_stats', '_MYREFERER_STATS_QUERY_PAGES', '100', '_MYREFERER_STATS_QUERY_PAGES_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (28, 'myref_referer_stats', '_MYREFERER_STATS_REFERER', '100', '_MYREFERER_STATS_REFERER_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (29, 'myref_referer_pages_stats', '_MYREFERER_STATS_REFERER_PAGES', '100', '_MYREFERER_STATS_REFERER_PAGES_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (30, 'myref_robots_stats', '_MYREFERER_STATS_ROBOTS', '100', '_MYREFERER_STATS_ROBOTS_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (31, 'myref_robots_pages_stats', '_MYREFERER_STATS_ROBOTS_PAGES', '100', '_MYREFERER_STATS_ROBOTS_PAGES_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (32, 'myref_users_stats', '_MYREFERER_STATS_USERS', '100', '_MYREFERER_STATS_USERS_DSC', 'select', 'int', 500);
INSERT INTO `arma_myref_config` VALUES (33, 'myref_users_pages_stats', '_MYREFERER_STATS_USERS_PAGES', '100', '_MYREFERER_STATS_USERS_PAGES_DSC', 'select', 'int', 500);


docs/changelog.txt
<h5>1.32 Alpha 1   [WORK IN PROGRESS - NOT RELEASED]</h5>  Dev: XOOPS 2.5.11, PHP 7.4.7
<hr>
- namespaces (mamba)
- class autoloading (mamba)
- added Feedback (mamba)
- added Migrate (mamba)
- added Stats (mamba)
- renamed 'PHP_SELF' to 'SCRIPT_NAME' (mamba)

- moved include/config.php to /config/config.ph (mamba)
- __NAMESPACE__ in Helper.php (mamba)


- reverted in_array() true check (mamba)
- reversed to while (list()) (mamba)

WORK IN PROGRESS

<h5>2.xx Beta 1  [2017-07-20]</h5>
<hr>

- XOOPS 2.5.9, PHP 7 (mamba)
- removed @version (mamba)
- PSR-2 code cosmetics (mamba)
- addNavigation(basename(__FILE__))  (mamba)
- dirname(__FILE__) to __DIR__ (mamba)
- replaced < br /> with < br > (mamba)
- Unnecessary double quotes (mamba)
- reference mismatch (mamba)
- Hardening to type safe (mamba)
- update Help, .html to .tpl (mamba)
- changed _handler to Handler (mamba)
- removed '/' from void HTML elements (&lt;br&gt;, &lt;img&gt;, etc) (mamba)
- HTML 5: removed 'selected' and 'checked' (mamba)
- updated file headers for consistency (mamba)
- renamed several $indexAdmin with $adminObject for consistency (mamba)
- Updates for PayPal changes (zyspec)
- added Contributing.md (mamba)
- moved functions to Utility class (mamba)
- include_once to require_once (mamba)
- php-cs-fixer updates (mamba)
- short array syntax (mamba)
- Yoda notation (mamba)
- started conversion to XMF (mamba)
    &nbsp;&nbsp;- updated addInfoBoxLine entries (mamba)
    &nbsp;&nbsp;- displayNavigation(basename(__FILE__))  (mamba)
    &nbsp;&nbsp;- changed to ->displayButton('left') (mamba)


- added debug_backtrace (mamba)

- replaced $xoopsGTicket with $XoopsSecurity (Mamba)
- added tokens to forms in templates (mamba)
- added 'true' for tokens in new forms (mamba)


- added install checks for PHP 5.5 and XOOPS 2.5.8 (mamba)
- added auto-creation of upload folders (mamba)


- moved all images, CSS, and JS files to /assets (mamba)
- renamed .html Smarty templates to .tpl (mamba)

- added SCANDIR_SORT_NONE to scandir()


- full XMF Conversion (mamba)
- added namespaces (mamba)
- with PSR-4 Autoloading (mamba)

- replaced each() with foreach() for PHP 7.2 (mamba)


Zyspec changes
-------------------------

- added breadcrumb display to top of index page
- added display of number of offline/online/waiting quotes in Admin index
- added significant security improvements for all vars passed from forms
- added module search capability
- added support for XOOPS Tag module
- added date quote was added to database
- changed to allow HTML in quote entry (XoopsRequest::getText) in admin
- changed install/update/unistall scripts
- changed hard coded language strings from templates to language files
- changed to improved index.tpl rendering
- changed to improved PSRx naming nomenclature
- changed phpDocumentor copyright, license, & link elements
- changed to clean up phpdocumentor comments
- fixed typo in block text length calculation
- fixed bug where 'item_tag' wasn't correctly assigned to template in /class/quotes.php
- fixed bug in install script PHP version checking
- fixed bug in install script XOOPS version checking
- fixed bug where language file wasn't being loaded in install script
- fixed typo in ./docs/licencs file name to license.txt
- general code cleanup
- removed deprecated cleanVarsRandomquote, replaced with XoopsRequest
- removed '/' from void HTML elements (&lt;br&gt;, &lt;img&gt;, etc)
- removed need for ./include/config.php
- removed need for ./include/action.module.php
- removed @version from phpDocumentor form page-level blocks
- removed unused admin templates
- fixed PHP version check (mamba)
- fixed uninstall (mamba)




08-01-08 Simon Tuck <stu@rtpartner.ch>, Rueegg Tuck Partner

	* BUGFIX: static typoscript directory missing from extension
	* CHANGED: Moved static typoscript to ext_typoscript_setup.txt
	* BUGFIX: Fixed debug library, tpl and function to avoid recursion
	* BUGFIX: prefilter dots now recognizes literals
	* BUGFIX: modifier format gets an instance of tslib_cobj
	* NEW: Additional params forceType, collapse and description added to modifier debug
	* NEW: plugins object, image, data, typoscript, typolink, LLL and format now use $smarty->trigger_error
	* BUGFIX: Rewrite of function get_debug_info.
	* BUGFIX: Modified output from dBug library to work within the context of TYPO3 and Smarty
	* BUGFIX: block function typoscript now calls cObjGet instead of cObjGetSingle
	* CHANGED: Moved default TypoScript settings to ext_typoscript files instead of using static templates
	* CHANGED: Rewrite of TCA helper functions for plugins

07-10-09 Simon Tuck <stu@rtpartner.ch>, Rueegg Tuck Partner

	* Complete rewrite of the extension
	* Added integration with the pi_base scenario as well as the lib/div scenario
	* Cleaned up the custom plugins
	* Added a debug console


admin/config.php
<?php declare(strict_types=1);

/**
 * XOOPS - PHP Content Management System
 * Copyright (c) 2004 <https://xoops.org>
 *
 * Module: myreferer 2.0
 * Licence : GPL
 * Authors :
 *           - solo (www.wolfpackclan.com/wolfactory)
 *            - DuGris (www.dugris.info)
 */

use XoopsModules\Myreferer\Config;
use XoopsModules\Myreferer\Helper;

require_once __DIR__ . '/admin_header.php';

xoops_cp_header();
$adminObject = \Xmf\Module\Admin::getInstance();
$adminObject->displayNavigation(basename(__FILE__));

// Activate php debug mode
error_reporting(E_ALL ^ E_NOTICE);

$op = 'edit';
if (isset($_POST)) {
    foreach ($_POST as $k => $v) {
        ${$k} = $v;
    }
}
if (isset($_GET['op'])) {
    $op = trim($_GET['op']);
}

if (isset($_GET['confcat_id'])) {
    $confcat_id = (int)$_GET['confcat_id'];
}

$helper        = Helper::getInstance();
$configHandler = $helper->getHandler('Config');

switch ($op) {
    case 'save':
        if (!$GLOBALS['xoopsSecurity']->check()) {
            redirect_header(basename(__FILE__), 3, implode('<br>', $GLOBALS['xoopsSecurity']->getErrors()));
        }
        $count = count($conf_ids);
        if ($count > 0) {
            for ($i = 0; $i < $count; $i++) {
                $config = new Config($conf_ids[$i]);

                $new_value = ${$config->getVar('conf_name')};

                $config->setConfValueForInput($new_value);

                if (!$configHandler->insert($config)) {
                    redirect_header(basename(__FILE__), 3, implode('<br>', $config->getErrors()));
                }
            }
        }
        redirect_header('index.php', 3, _MD_MYREFERER_DBUPDATE);
        break;
    case 'edit':
    default:
        // Utility::getAdminMenu(2, _MD_MYREFERER_CONFIG);

        $criteria = new CriteriaCompo();
        $criteria->add(new Criteria('conf_formtype', 'hidden', '!='));
        $criteria->setSort($sort);
        $criteria->setOrder($order);

        $config    = $configHandler->getAll($criteria);
        $confcount = count($config);
        myreferer_EditConfig($config);
        break;
}
require_once __DIR__ . '/admin_footer.php';

/**
 * @param $config
 */
function myreferer_EditConfig($config)
{
    xoops_load('XoopsThemeForm');

    $form = new XoopsThemeForm(_MD_MYREFERER_CONFIG, 'pref_form', basename(__FILE__), 'post', true);

    $helper = Helper::getInstance();

    $configHandler = $helper->getHandler('Config');

    foreach ($config as $i => $configObj) {
        $title = !defined($config[$i]->getVar('conf_desc')) || '' == constant($config[$i]->getVar('conf_desc')) ? constant($config[$i]->getVar('conf_title')) : constant($config[$i]->getVar('conf_title')) . '<br><br><span style="font-weight:normal;">'
                                                                                                                                                                . constant($config[$i]->getVar('conf_desc')) . '</span>';

        switch ($config[$i]->getVar('conf_formtype')) {
            case 'insertBreak':
                $form->insertBreak('', 'null');
                $form->insertBreak('<span style="font-weight: bold; text-align: center;">' . $title . '</span>', 'bg3');
                break;
            case 'textarea':
                $myts = \MyTextSanitizer::getInstance();
                if ('array' === $config[$i]->getVar('conf_valuetype')) {
                    // this is exceptional.. only when value type is arrayneed a smarter way for this

                    $ele = '' != $config[$i]->getVar('conf_value') ? new XoopsFormTextArea($title, $config[$i]->getVar('conf_name'), $myts->htmlSpecialChars(implode('|', $config[$i]->getConfValueForOutput())), 5, 50) : new XoopsFormTextArea($title,
                                                                                                                                                                                                                                                 $config[$i]->getVar('conf_name'),
                                                                                                                                                                                                                                                 '', 5,
                                                                                                                                                                                                                                                 50);
                } else {
                    $ele = new XoopsFormTextArea($title, $config[$i]->getVar('conf_name'), $myts->htmlSpecialChars($config[$i]->getConfValueForOutput()), 5, 50);
                }
                break;
            case 'select':
                if ('save_group' === $config[$i]->getVar('conf_name')) {
                    $ele = new XoopsFormSelectGroup($title, $config[$i]->getVar('conf_name'), false, $config[$i]->getConfValueForOutput(), 5, true);
                } else {
                    $ele = new XoopsFormSelect($title, $config[$i]->getVar('conf_name'), $config[$i]->getConfValueForOutput());

                    $options = [_MYREFERER_STATS_ALL => 0];

                    for ($j = 25; $j <= 200; $j += 25) {
                        $options[_MYREFERER_STATS_TOP . ' ' . $j] = $j;
                    }

                    foreach ($options as $optkey => $optval) {
                        $ele->addOption($optval, $optkey);
                    }
                }
                break;
            case 'select_multi':
                $ele     = new XoopsFormSelect($title, $config[$i]->getVar('conf_name'), $config[$i]->getConfValueForOutput(), 5, true);
                $options = $configHandler->getConfigOptions(new Criteria('conf_id', $config[$i]->getVar('conf_id')));
                $opcount = count($options);
                for ($j = 0; $j < $opcount; $j++) {
                    $optval = defined($options[$j]->getVar('confop_value')) ? constant($options[$j]->getVar('confop_value')) : $options[$j]->getVar('confop_value');

                    $optkey = defined($options[$j]->getVar('confop_name')) ? constant($options[$j]->getVar('confop_name')) : $options[$j]->getVar('confop_name');

                    $ele->addOption($optval, $optkey);
                }
                break;
            case 'yesno':
                $ele = new XoopsFormRadioYN($title, $config[$i]->getVar('conf_name'), $config[$i]->getConfValueForOutput(), _YES, _NO);
                break;
            case 'textbox':
            default:
                $myts = \MyTextSanitizer::getInstance();
                $ele  = new XoopsFormText($title, $config[$i]->getVar('conf_name'), 50, 255, $myts->htmlSpecialChars($config[$i]->getConfValueForOutput()));
                break;
        }

        $hidden = new XoopsFormHidden('conf_ids[]', $config[$i]->getVar('conf_id'));

        $form->addElement($ele);

        $form->addElement($hidden);

        unset($ele);

        unset($hidden);
    }

    $form->addElement(new XoopsFormHidden('op', 'save'));

    $form->addElement(new XoopsFormButton('', 'button', _GO, 'submit'));

    $form->display();
}
